pyvo/dal/sia.py

```diff
@@ -345,7 +345,7 @@ class SIAQuery(DALQuery):
     """
 
     def __init__(
-            self, baseurl, pos=None, size=None, format=None, intersect=None,
+            self, session, baseurl, pos=None, size=None, format=None, intersect=None,
```

Comment:
Have to be careful to break the API signature this way, if we go ahead a big warning needs to be made, maybe even some checking to raise exceptions early for anyone using the old API with a new version.

Reply:
Right, I did it this way so that the breaking changes would actually help me find the chain of objects, which isn't exactly straightforward to me with all the mixins.  But I think we could pass it in as a kwarg instead with maybe the default just making its own non authenticated session.  (Although that's a lot harder than 'expected 2 got 1' errors :) ).  I hope that we can have a backward compatible API for this.

Reply:
I was also possibly thinking about just having one new object that combines session and url, since everything has both now, and make some kind of context object that could have other things, and allow us to add more things in a somewhat global manner, without having to change all the signatures.

Reply:
IMO it would be much better to provide session as a separate keyword argument.

Reply:
Okay sounds like we're all in agreement on this one then!

Reply:
I still have some reservations about exposing the session in the constructor as a user credential holder (for this and the other service classes):
1. At its simplest, users might not want to be exposed to the underlying `requests` library. They might want to instantiate the server and use their credentials (be it a path to a cert file, a user/password, a cookie or token) execute queries and get back tables with the results. @funbaker you've made a comment that `pyvo` is an API and `astroquery` is a user interface. At the CADC we are already offering user catalogues (cloud databases) and `pyvo` would be the tool to access them whereas I see `astroquery` as the tool to access data. IMO both of them should be user interfaces. BTW, power users have access to the session and can customize it if they wish.
2. I'm confused about the role of authentication methods in capabilities if the library does not try to match them with user credentials? Are those for meant for user or machine consumption? Are users supposed to read the capabilities document and create the appropriate session? Not all the astronomers that I work with would be happy to do that in their scripts. Again, general vs power users. Library should make it as simple as possible for general users while allowing power users to take matters into their own hands if they want to.
3. Keeping the session internal allows the class to implement optimizations or enhancements. More robust session that does automatic retries on transient errors for ex.
I see the advantage of not re-inventing the wheel for capturing user credentials when this is already available in a stable, well documented package and for me it makes sense. At the same time, I'm also trying to look at the API from the point of view of a general astronomer user that just wants simple and fast access to data.

Reply:
The problem is that there are many non-standard authentication methods in the VO.
It's easy to implement e.g. HTTP Basic Auth in a generic way, but things like Cookie authentication is service specific (different key, different hash algorythms, etc)

> I still have some reservations about exposing the session in the constructor as a user credential holder

Exposing as a keyword param with a default value removes the limitation of the user having to deal with it if not needed.

> I'm confused about the role of authentication methods in capabilities if the library does not try to match them with user credentials?

Pinging @msdemlei for detailed explanation. AFAIK both cases apply, hence the `describe()` methods.

Reply:
Agree with @funbaker 

As to the capabilities and how they are matched up with user credentials, before this PR, the securitymethods were parsed but not used (other than printing them out in describe) as far as I know.  This is the first usage of these classes and parts of the capabilities.

The new code in the PR has the ability to look at the capabilities of the service, and what creds it has been given by the user, and matching that up to see if anything works.  [0] is a description of the steps there.

The steps in [0] match up the methods against what creds we have for that authsession.  But past the method (cookie, basic, cert), it doesn't give any hint as to which creds of that type it is looking for.  For example, it may say tls-with-certificate and pass a client certificate, but it doesn't say that the certificate needs to come from a particular authority or have a special field.

Let's go through a somewhat complex example that points out some of the trickiness.  Say you have two certificates now, one for Gaia, and one for CADC.  And you want to use the Gaia cert to auth with Gaia, and CADC cert to auth with CADC.  From the capabilities of each service, they could both say tls-with-certificate.  But they don't say anything about how to pick which cert they are talking about (there's no subject line, or signing authority, etc).

If you pick the wrong cert, you are basically disclosing it to the wrong service, which is exposing your creds (hopefully they don't log it :smile: ).  Similar cases exist for username/password.

The underlying calling pattern is due to the ambiguous nature of which creds go with what services, and for security, it's dangerous to guess. Without the knowledge to figure out which creds go with what service from the capabilities, we have to rely on the user to tell us by which creds should be used for what services.  

We can add in more helpers to set this up in an easier way, but I was also trying to make this a change that people could use, but wouldn't possibly break anyone that wasn't using it.  I guess some of this gets into how "first-class" do you want an extension to be in the API?

[0]
1) Make an authsession and attach your creds to it by adding providers.  This is done by putting in the URI of the security method and the requests class to handle that URI. 
2) Attach the authsession to a service, it gets the capabilities of the service using pyvo and requests to get the XML document, and it parses out the security methods.
3) That is passed into the authurls class, which iterates over the interfaces in the capabilities, and logs their URLs, base or full, and their security method URI.
4) Now the user makes a request.  The session gets the URL of the request, and tries to match it up against URLs in the interfaces mentioned in step 2, taking into account full and base paths.  It comes up with a list of the auth methods that the server will accept.
5) The intersection of the server methods is taken with the creds that have been put in the jar.  Then we pick one and use it for auth for that particular request.